### PR TITLE
WSTEAM1-768 Live Label Compound Conversion

### PR DESCRIPTION
**Pull request by HarveyPeachey**
Resolves JIRA [WSTEAM1-768](https://jira.dev.bbc.co.uk/browse/WSTEAM1-768)

Overall changes
======
Converts the `LiveLabel` component into a compound component to help support work with Live pages in Next.js. This compound component is then used to create the `LiveLabelPromo` applied compound component to keep the existing component functionality for all pages where the label is used in Promos.

Code changes
======

- Converts `LiveLabel` into a compound component
- Updates existing snapshots which moves the width and height to bottom of emotion css classes

Testing
======
1. A quick visual regression test

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/components/LiveLabelPromo/index.tsx
Comment (karinathomasbbc) [inline src/app/components/LiveLabelPromo/index.tsx]: I am wondering if we could agree on a better name for this - it's not a Promo, it's just the Pulse & Label together... (I can't think of a good name right now, but I know that LiveLabel kinda implies it's a label only, and not a pulse 🤦‍♀️ ). `LiveComponent`? Nope! `LiveIndicator`?

**Comment by karinathomasbbc** [comment]
FY of other reviewers, @HarveyPeachey and I have been discussing this offline. I wondered whether this approach was actually necessary, as we had hoped to be able to just drop the component onto the LivePage without needing to make any changes. However, it appears that there is slightly different styling required at different breakpoints, as well as the live label's positioning in the live page heading. We don't want to have to add a lot of complexity to the component to support this behaviour, so perhaps the compound component is the way to go... 

@amoore108 and @andrewscfc, I was wondering if you had any thoughts on this?

**Comment by karinathomasbbc** [comment]
I wonder if it's worth spinning up another branch from the WSTEAMA-927-fix-live-pulse-pixelation branch, but just using the LiveLabel component as is on the LivePage, to see where the gaps are?

File: src/app/components/LiveLabelPromo/index.tsx
Comment (HarveyPeachey) [reply src/app/components/LiveLabelPromo/index.tsx]: Yeah just came up with a quick name. I think the parent compound component should be called `LiveIndicator`, but the applied one is used on promos, hence why I named itthat way, `LiveIndicatorForPromos` maybe? but that looks weird as well

Comment (karinathomasbbc) [reply src/app/components/LiveLabelPromo/index.tsx]: Or perhaps `LivePromoIndicator`?

**Comment by HarveyPeachey** [comment]
> I wonder if it's worth spinning up another branch from the WSTEAMA-927-fix-live-pulse-pixelation branch, but just using the LiveLabel component as is on the LivePage, to see where the gaps are?

Here ya go https://github.com/bbc/simorgh/pull/11278, I've literally just placed it in. The PR has a before and after, I've highlighted the text as it's invisible if I didn't

File: src/app/components/LiveLabelPromo/index.tsx
Comment (karinathomasbbc) [reply src/app/components/LiveLabelPromo/index.tsx]: Actually, thinking about this a bit more - let's not even add the concept of Promo at all. I think the parent compound component should be called `LiveIndicator`, and the other components called `LivePulse` and `LiveText`. Our Homepages will use the `LiveIndicator`, while the LivePage will use the `LivePulse` + `LiveText` components with additional styling applied.

**Comment by andrewscfc** [comment]
> FY of other reviewers, @HarveyPeachey and I have been discussing this offline. I wondered whether this approach was actually necessary, as we had hoped to be able to just drop the component onto the LivePage without needing to make any changes. However, it appears that there is slightly different styling required at different breakpoints, as well as the live label's positioning in the live page heading. We don't want to have to add a lot of complexity to the component to support this behaviour, so perhaps the compound component is the way to go...
> 
> @amoore108 and @andrewscfc, I was wondering if you had any thoughts on this?

This PR doesn't include the usage within the live page as I understand it, so this is just preparing the ground for it's usage in that context? The 'applied compound component' is terminology I coined to try and give a name to a component that has been customised for use on a specific page and is not reusable. It sounds like you have:
- made `LiveLabel` a compound component to allow customised styling
- created `LiveLabelComponent` a 'complex component' that is reused extensively

The applied compound component will come when you use the `LiveLabel` compound component on the live page with custom styling.

It feels like a sensible use of the pattern to me as the alternative is to have a live label component with configuration that varies it's layout which violates the 'open-closed' principle.

Here's the docs again on the compound component stuff so you know what I'm referencing: https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/docs/Coding-Standards/React-Components.stories.mdx#L186

File: src/app/components/LiveLabelPromo/index.tsx
Comment (HarveyPeachey) [reply src/app/components/LiveLabelPromo/index.tsx]: So the [parent compound component](https://github.com/bbc/simorgh/blob/5e7e2d95ec5bc809be12bfe8b4221c9e657ecd90/src/app/components/LiveLabel/index.tsx#L7-L22), acts as an entry point for importing `Pulse` and `Text`, it does nothing with the layout of those components, and handles any state needed for them. Could it get confusing if we turn them into separate components?

Are you suggesting something like this?:
```
- components
-- LiveIndicator (Used throughout)
-- LivePulse
-- LiveText
```
Instead of:
```
- components
-- LiveIndicator
--- Pulse
--- Text
-- LiveLabelPromo (Used throughout)
```

Comment (karinathomasbbc) [reply src/app/components/LiveLabelPromo/index.tsx]: If the first option works, I would prefer that one (although would welcome a second opinion if anyone had one? And I agree about the potential complications of having lots of components) I am not sure what this does for your implementation - e.g. do you still need the LiveLabelContext?

**Comment by karinathomasbbc** [comment]
> Here's the docs again on the compound component stuff so you know what I'm referencing:

Here's a nicer view: https://latest--5d28eb3fe163f6002046d6fa.chromatic.com/?path=/story/coding-standards-react-components--page#primitive-components-compound-components-and-complex-components

File: src/app/components/LiveLabelPromo/index.tsx
Comment (HarveyPeachey) [reply src/app/components/LiveLabelPromo/index.tsx]: I think there's a few gaps in our component standards around how these compound components are organised. 

For example in the OptimoPromos, we add the [service to the compound components context](https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/OptimoPromos/index.jsx#L24). Should we do that in this instance? Confusingly the OptimoPromos also imports the [service context](https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/OptimoPromos/Content/index.jsx#L10-L12) in child components as well. It's also confusing as it houses components which [live somewhere separately or are used outside of the OptimoPromos](https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/OptimoPromos/index.jsx#L43-L46).

Then on the need for `LiveLabelContext`, all compound patterns we've used so far have context involved, [OptimoPromos](https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/OptimoPromos/index.jsx#L28-L30) and [Promo](https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/Promo/index.jsx#L38), does that mean we ditch the context when context isn't needed? Or do we need context because that's where we pass down other contexts like request and service for example? Also does it get confusing for devs if they a see a compound component with no context boilerplate? The [article](https://kentcdodds.com/blog/compound-components-with-react-hooks) we link to for this pattern always has context in its examples, and they even make it mandatory to have the parent component encapsulating it all.

Comment (amoore108) [reply src/app/components/LiveLabelPromo/index.tsx]: Yea I don't think we should be passing `service` from ServiceContext down through _another_ context. ServiceContext is available throughout the entire application in any renderable component, so I don't see the need in passing it as a prop.

Similarly, as you say we shouldn't reference external components as part of a compound component. Fairly dangerous as removal of those external components would then break the compound component that references them.

Comment (amoore108) [reply src/app/components/LiveLabelPromo/index.tsx]: I'd say for the `LiveLabelContext`, or any context, to only use it if its necessary and helps reduce the amount of state you're passing around.

If `LiveLabelContext` isn't used then I'd remove it and add it only when we need to.

A lot of the examples of the compound pattern seem to be more interactive applications where state is uplifted to the top and then passed down in the context provider. Since we have access to all our context providers anyway, then it doesn't seem needed here unless we need to access props passed to LiveLabel in its sub-components?

Comment (karinathomasbbc) [reply src/app/components/LiveLabelPromo/index.tsx]: Just a quick question/clarification on 
> we shouldn't reference external components as part of a compound component. Fairly dangerous as removal of those external components would then break the compound component that references them.

Do you mean that the compound - in this case - LiveIndicator - should not reference external / sibling components? We would end up with broken compound components if someone were to remove primitive components such as `Image`, `Timestamp` etc, which live in the src/app/components directory. So I am not sure if we need to be concerned about that in this case? I think any of us would check references to components before making changes. We also have visual snapshot tests (Chromatic) to make sure that our components are still rendering as expected.

Or do you think that the second option would avoid this situation, even if we did end up with some duplication?

I think we will need to be really clear in the usage docs for all of these components, so that people know which component to use, when, and in which context (not the `context` we're speaking about ☝️ 😅 )

Comment (amoore108) [reply src/app/components/LiveLabelPromo/index.tsx]: I mean in Harveys example of the OptimoPromo https://github.com/bbc/simorgh/blob/6fe88689a7cb76f8afe67bed7338ee9da1c16297/src/app/legacy/components/OptimoPromos/index.jsx#L43-L46, I feel we shouldn't be exposing `Promo.Timestamp = Timestamp;` as anyone consuming the `Promo.Timestamp` would rightfully assume that it comes from `OptimoPromo` since its exported from there, but in reality it isn't.

Using external primitive components within compound components and its siblings is fine, but exposing an external component as a compound component sibling isn't imo.

Comment (amoore108) [reply src/app/components/LiveLabelPromo/index.tsx]: In that example, we could have created a local `Timestamp` component and imported the Psammead Timestamp into that and then exported it as `Promo.Timestamp = Timestamp`. 

It would be a very light/thin wrapper around the Psammead Timestamp component, but its clearer what its doing in that case. Whereas atm with those OptimoPromo exports, we are locked in to the Psammead Timestamp component.

Comment (HarveyPeachey) [reply src/app/components/LiveLabelPromo/index.tsx]: > In that example, we could have created a local `Timestamp` component and imported the Psammead Timestamp into that and then exported it as `Promo.Timestamp = Timestamp`.

Sounds like this could be done as part of the impactful promo work if OptimoPromos is used

Comment (HarveyPeachey) [reply src/app/components/LiveLabelPromo/index.tsx]: Anyway @karinathomasbbc, I've moved the layout logic into the parent compound component. I've decided not to rename it in this PR, as I didn't want to make it too bloated to review as there are a lot of references to live label in documentation and the code, and it allows us to move this along a bit faster. We can open up a new ticket if we've got a strong opinion on changing it. 

I've kept the pulse and text within the parent compound component, as I think co-location is better, but I've kept the LiveLabel.Pulse and LiveLabel.Text exports, meaning we can import LiveLabel directly and reference those child components on the live pages. 

I've kept a note of some of the things we've discussed so if we do another component standards session to clarify some thing around components, we can bring them up in the meeting and update our docs

Comment (karinathomasbbc) [reply src/app/components/LiveLabelPromo/index.tsx]: Sounds good, thank you!

File: src/app/components/LiveLabel/Pulse/index.styles.tsx
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/Pulse/index.styles.tsx:1]: While you're in here, can we rename the file to .ts instead of .tsx - there is no React stuff in here 😄

File: src/app/components/LiveLabel/Text/index.styles.tsx
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/Text/index.styles.tsx:1]: As above, rename to .ts instead of .tsx

File: src/app/components/LiveLabel/__snapshots__/index.test.tsx.snap
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/__snapshots__/index.test.tsx.snap]: Has the height/width actually changed or has it moved location? Will this affect the visuals? (we might need to do a couple of quick a11y checks e.g. CSS disabled, to ensure that it's still rendering as expected)

**Review by amoore108 — APPROVED** [review]
Great work here!

File: src/app/components/LiveLabel/index.stories.tsx
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/index.stories.tsx]: Shall we remove 'Promo' from the name?
```suggestion
  title: 'New Components/Live Label',
```

File: src/app/components/LiveLabel/index.styles.tsx
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/index.styles.tsx:1]: Rename to .ts

File: src/app/components/LiveLabel/types.ts
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/types.ts]: Perhaps this is overkill, but could LiveLabel props extend TextProps, so that we don't need to duplicate the types?

```
export interface LiveLabelProps extends TextProps {
  className?: string
}
```

File: src/app/components/LiveLabel/__snapshots__/index.test.tsx.snap
Comment (HarveyPeachey) [reply src/app/components/LiveLabel/__snapshots__/index.test.tsx.snap]: I've mirrored the last change in the pixelation fix PR Lou is working on, that's why this has changed https://github.com/bbc/simorgh/pull/11212/commits/4ef35aba6e56c0853866f8b3ed82cb58e3bf3d77

Comment (karinathomasbbc) [reply src/app/components/LiveLabel/__snapshots__/index.test.tsx.snap]: Great, thank you. This is always tricky when you're working off a branch which is still being modified.

File: src/app/components/LiveLabel/types.ts
Comment (HarveyPeachey) [reply src/app/components/LiveLabel/types.ts]: All done

File: src/app/components/LiveLabel/types.ts
Comment (karinathomasbbc) [inline src/app/components/LiveLabel/types.ts:9]: **nitpick** I assume the order of this doesn't matter i.e. TextProps don't need to be declared higher in the file before being used?

**Review by karinathomasbbc — APPROVED** [review]
Thanks so much for this, looks good!

I am assuming we will need to wait until the base branch has been merged? Also, are we happy that we don't need to run any more a11y checks based on these changes?

File: src/app/components/LiveLabel/types.ts
Comment (HarveyPeachey) [reply src/app/components/LiveLabel/types.ts:9]: Hmm yeah I'm not sure, I'll put it after just in case. I'm not getting any vscode compiler error hints saying anything is wrong

**Comment by HarveyPeachey** [comment]
> I am assuming we will need to wait until the base branch has been merged? Also, are we happy that we don't need to run any more a11y checks based on these changes?

Yeah, I think that's best. I don't _think_ we do, as it doesn't change anything semantically, or tag order so it should be fine

File: src/app/components/LiveLabel/types.ts
Comment (karinathomasbbc) [reply src/app/components/LiveLabel/types.ts:9]: Yeah I think we leave it, looks like it's able to cater for it regardless of its position in the file

**Pull request merged**
